internal/dag/gatewayapi_processor.go

```diff
@@ -230,24 +230,98 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
 	return strings.ToLower(certificateRef.Kind) == "secret" && strings.ToLower(certificateRef.Group) == "core"
 }
 
-func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname) ([]string, []error) {
-	// Determine the hosts on the route, if no hosts
-	// are defined, then set to "*".
-	var hosts []string
+// computeHosts validates the hostnames for a HTTPRoute as well as validating
+// that the hostname on the HTTPRoute matches what is optionally defined on the
+// listener.hostname.
+func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname, listenerHostname *gatewayapi_v1alpha1.Hostname) (map[string]struct{}, []error) {
+
+	hosts := make(map[string]struct{})
 	var errors []error
-	if len(hostnames) == 0 {
-		hosts = append(hosts, "*")
+
+	// Determine the hosts on the hostnames, if no hosts
+	// are defined, then set to "*". If the listenerHostname is defined,
+	// then the route must match the Gateway hostname.
+	if len(hostnames) == 0 && listenerHostname == nil {
+		hosts["*"] = struct{}{}
 		return hosts, nil
 	}
 
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(hostnames) == 0 {
+				hosts[string(*listenerHostname)] = struct{}{}
+				return hosts, nil
+			}
+		}
+	}
+
 	for _, host := range hostnames {
 
 		hostname := string(host)
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
+			hosts[hostname] = struct{}{}
+			continue
+		}
+
+		// Validate the hostname.
 		if err := validHostName(hostname); err != nil {
 			errors = append(errors, err)
 			continue
 		}
-		hosts = append(hosts, string(host))
+
+		if listenerHostname != nil {
+			const wildcardString string = "[a-z0-9]([-a-z0-9]*[a-z0-9])*."
+			lhn := string(*listenerHostname)
+
+			// A "*" hostname matches anything.
+			if lhn == "*" {
+				hosts[hostname] = struct{}{}
+				continue
+			} else if lhn == hostname {
+				// If the listener.hostname matches then no need to
+				// do any other validation.
+				hosts[hostname] = struct{}{}
+				continue
+			} else if strings.Contains(lhn, "*") {
+
+				// Build a regex to match the hostname against the wildcard listener hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(lhn, "*", wildcardString, 1) + "$")
```

Comment:
nit: I think this could be simplified to not need a regex match? If the listener hostname is a wildcard (`*.foo.com`) just chop the first DNS label from that and the route hostname and those suffixes must be equal

not super critical but makes it harder to make a mistake in case we get our regex wrong etc.

Reply:
nope I like it!